File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>

using namespace std;

int answer = 0;

void dfs(int currentPoint, vector<vector<int>>& d, vector<bool> visited, int dCount)
{
    if (dCount > answer)
        answer = dCount;
    
    for(int i = 0; i < d.size(); ++i)
    {
        if (!visited[i] && d[i][0] <= currentPoint)
        {
            visited[i] = true;
            dfs(currentPoint - d[i][1], d, visited, dCount + 1);
            visited[i] = false;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    
    dfs(k, dungeons, vector<bool> (dungeons.size(), false), 0);
    
    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    public static int answer = -1;

    public int solution(int k, int[][] dungeons) {
        dfs(0, k, new boolean[dungeons.length], dungeons);
        return answer;
    }

    public void dfs(int count, int k, boolean[] visited, int[][] dungeons) {
        if (count > answer) {
            answer = count;
        }
        for (int i = 0; i < dungeons.length; i++) {
            if (visited[i] || k < dungeons[i][0]) continue;

            visited[i] = true;
            dfs(count + 1, k - dungeons[i][1], visited, dungeons);
            visited[i] = false;
        }
    }
}